### PR TITLE
E2Es for Recent Episodes Component (Radio and TV)

### PR DESCRIPTION
**Pull request by LilyL0u**
**Overall change:**

These E2Es run against both OD Radio and TV, canonical and amp. They check the toggle is enabled for the service, and the value of the toggle (max number allowed to show). If the toggle is enabled it checks for the present of the component on the page. It then also checks the number of items is expected given the maximum value in the toggles, the number of episodes in the data, and the number of those episodes that are available (do not have versions missing). If the number of episodes is 1, it checks a different locator because it is not in a list when there is only 1 item.

Currently the od tv code is using class locators to identify parts of the component, but when we add the data-e2e indicators for the tv component, this will use those.

All the relevant services have their URLs enabled in the settings. (Including Zhongwen, which does not fail as it goes directly to the correct URL)

I have made it so this will not run on local, as it will fail: the toggle environment for local is test, but the recent episodes component does not show on these local pages until all the fixture data is updated. I will do this at some point.

---

- [x] I have assigned myself to this PR and the corresponding issues
- [ ] I have added the `cross-team` label to this PR if it requires visibility across World Service teams
- [x] I have assigned this PR to the Simorgh project

**Testing:**

These pass on test
I have made it so this will not run on local, as it will fail: the toggle environment for local is test, but the recent episodes component does not show on these local pages.

### CODE REVIEW

**Review by SetJackal — APPROVED** [review]

File: cypress/integration/pages/onDemandRadio/tests.js [hunk 1]
```diff
@@ -53,5 +55,97 @@ export default ({ service, pageType, variant, isAmp }) => {
...
+                    i < numberOfEpisodesInDataUnderLimit;
+                    i += 1
+                  ) {
+                    if (
+                      body.relatedContent.groups[0].promos[i].media.versions
+                        .length > 0
+                    ) {
+                      countAvailableEpisodes += 1;
+                    }
+                  }
```
Comment (jroebu14) [inline cypress/integration/pages/onDemandRadio/tests.js]:
```suggestion
                  const countAvailableEpisodes  = body.relatedContent.groups[0].promos.filter(({ media }) => media.versions.length).slice(0, recentEpisodesMaxNumber).length
```

Comment (jroebu14) [reply cypress/integration/pages/onDemandRadio/tests.js]: I think my suggestion above is actually your `expectedNumberOfEpisodes`

Comment (LilyL0u) [reply cypress/integration/pages/onDemandRadio/tests.js:110]: Nice! I need to remind myself about how filter works. Haven't seen it for a while. Will commit.

Comment (LilyL0u) [reply cypress/integration/pages/onDemandRadio/tests.js]: Aha, so I can also remove the part after this that uses the maxNumberOfEpisodes in the Math.min to get the expectedNumberOfEpisodes . I'll try it

Comment (LilyL0u) [reply cypress/integration/pages/onDemandRadio/tests.js]: Does this also mean I don't need to do 
const numberOfEpisodesinData =
                    body.relatedContent.groups[0].promos.length;
Because this method of counting episodes that have versions, would always give a number equal to or lower than the number of items in the data.

Comment (jroebu14) [reply cypress/integration/pages/onDemandRadio/tests.js]: yeh I think so - if you have linting set up you can change your `expectedNumberOfEpisodes` variable to my suggestions and then delete all the red squiggly code 🙂

Comment (LilyL0u) [reply cypress/integration/pages/onDemandRadio/tests.js]: I'm just doing it in vscode so I can remove other stuff that has become unecessary and watch it run. So far so good

Comment (jroebu14) [reply cypress/integration/pages/onDemandRadio/tests.js]: as a side note, with javascript array iteration methods, you should almost never have to write a for loop. I haven't written a for loop in years

https://www.w3schools.com/js/js_array_iteration.asp

I mostly use `map` and `filter`

File: cypress/integration/pages/liveRadio/testsForCanonicalOnly.js [hunk 2]
```diff
@@ -19,7 +20,7 @@ export const testsThatFollowSmokeTestConfigForCanonicalOnly = ({
 }) =>
   describe(`testsThatFollowSmokeTestConfigForCanonicalOnly for ${service} ${pageType}`, () => {
     beforeEach(() => {
-      cy.getToggles(service);
+      cy.getToggles(config[service].name);
```
Comment (RichardPK) [inline cypress/integration/pages/liveRadio/testsForCanonicalOnly.js]: Can changes in this file be broken out into their own PR? I'm not sure if they relate to recent episodes?

File: cypress/integration/pages/onDemandRadio/index.js [hunk 3]
```diff
@@ -11,7 +11,13 @@ const pageType = 'onDemandRadio';
 Object.keys(config)
   .filter(service => serviceHasPageType(service, pageType))
   .forEach(serviceId => {
-    const { variant, name: service } = config[serviceId];
+    // eslint-disable-next-line prefer-const
+    let { variant, name: service } = config[serviceId];
```
Comment (RichardPK) [inline cypress/integration/pages/onDemandRadio/index.js:15]: Am I right in thinking this reconstruction of the service is for accessing the services within `cypress/support/config/settings.js` when doing something like `config[service].name`?

Taking Zhongwen as an example - `config[service].name` accesses the config for `zhongwenTrad` and returns the name, "zhongwen" - which is the same as the `service` prop before you capitalise it and add the variant.

So for example, in `cypress/integration/pages/onDemandRadio/tests.js` does: 

```js
const language = appConfig[config[service].name][variant].lang;

// where service = 'zhongwenTrad'
```

have to access the config via `config[service].name` (which needs the service to be service + capitalised variant)

or could it just be:

```js
const language = appConfig[service][variant].lang;

// where service = 'zhongwen'
```

Basically - is there any instance where the `service` prop received in the test index does not equal `config[service].name`?

File: cypress/integration/pages/onDemandRadio/testsForCanonicalOnly.js [hunk 4]
```diff
@@ -1,65 +1,70 @@
...
+          cy.log(
+            `On Demand Radio Page configured for Radio Schedule? ${scheduleIsEnabled}`,
           );
-          if (Cypress.env('APP_ENV') === 'test') {
-            schedulePath += '?renderer_env=live';
-          }
 
-          cy.request(schedulePath).then(({ body: scheduleJson }) => {
-            const { schedules } = scheduleJson;
+          if (scheduleIsEnabled) {
```
Comment (RichardPK) [inline cypress/integration/pages/onDemandRadio/testsForCanonicalOnly.js:32]: As with my other comment - can these schedule test changes be broken into their own PR?

File: cypress/integration/pages/liveRadio/testsForCanonicalOnly.js (hunk 2 above)
Comment (RichardPK) [reply cypress/integration/pages/liveRadio/testsForCanonicalOnly.js]: Ah - see [my other comment](https://github.com/bbc/simorgh/pull/8597#discussion_r544299981) which relates to using `config[service].name)` instead of just `service`.

Is there a reason for switching?

File: cypress/integration/pages/onDemandRadio/index.js (hunk 3 above)
Comment (LilyL0u) [reply cypress/integration/pages/onDemandRadio/index.js:15]: changed

File: cypress/integration/pages/onDemandRadio/tests.js (hunk 1 above)
Comment (LilyL0u) [reply cypress/integration/pages/onDemandRadio/tests.js]: changed :)

File: cypress/integration/pages/liveRadio/testsForCanonicalOnly.js (hunk 2 above)
Comment (LilyL0u) [reply cypress/integration/pages/liveRadio/testsForCanonicalOnly.js]: Changed back

File: cypress/integration/pages/onDemandRadio/testsForCanonicalOnly.js (hunk 4 above)
Comment (LilyL0u) [reply cypress/integration/pages/onDemandRadio/testsForCanonicalOnly.js:32]: changed back

**Review by jroebu14 — APPROVED** [review]
Really nice!

File: cypress/integration/pages/onDemandRadio/index.js [hunk 5]
```diff
@@ -11,7 +11,9 @@ const pageType = 'onDemandRadio';
...
```
Comment (RichardPK) [inline cypress/integration/pages/onDemandRadio/index.js:15]: This can be changed back to a const now Lily 👍  (and remove the comment above)

File: cypress/integration/pages/onDemandRadio/tests.js [hunk 6]
```diff
@@ -53,5 +54,74 @@ export default ({ service, pageType, variant, isAmp }) => {
...
+      before(() => {
+        cy.getToggles(service);
+      });
+      describe('Recent Episodes component', () => {
+        it('should be displayed if the toggle is on, and shows the expected number of items (does not run on local)', function test() {
+          // I can enable tests running on local when I have updated all the fixture data to show the component
+          if (Cypress.env('APP_ENV') === 'local') {
+            cy.log('Does not run on local');
+          } else {
+            cy.fixture(`toggles/${service}.json`).then(toggles => {
```
Comment (RichardPK) [inline cypress/integration/pages/onDemandRadio/tests.js:67]: There's a loooot of nested conditionals here - maybe it's worth breaking this section into its own `recentEpisodesTests` test file? You may also then be able to reuse it for `onDemandTv`?

File: cypress/integration/pages/onDemandRadio/tests.js [hunk 7]
```diff
@@ -53,5 +54,74 @@ export default ({ service, pageType, variant, isAmp }) => {
...
+
+                    cy.get('[data-e2e=recent-episodes-list]').within(() => {
+                      cy.get('[data-e2e=recent-episodes-list-item]')
+                        .its('length')
+                        .should('eq', expectedNumberOfEpisodes);
+                    });
+                  }
+                  // If there is only one item, it is not in a list
+                  else if (expectedNumberOfEpisodes === 1) {
+                    cy.get("div[class*='css-1sel12u-Wrapper emzt7w80']").should(
```
Comment (RichardPK) [inline cypress/integration/pages/onDemandRadio/tests.js:107]: Some of these selectors are too specific - bits of this could well change: `1sel12u` & `emzt7w80` in particular.

There's some reasonable explanation of how different selectors work here https://www.w3schools.com/cssref/css_selectors.asp (scroll down to [attribute] to find the ones we're using a lot in cypress)

`class*=` which you're using here means 'select every element where the class contains this substring'. This is a super broad selector, but by passing in the entire className, it's actually super specific and makes the test quite brittle.

File: cypress/integration/pages/onDemandTV/tests.js [hunk 8]
```diff
@@ -39,5 +42,81 @@ export default ({ service, pageType, variant, isAmp }) => {
...
+
+                  cy.log(
+                    `Number of available episodes? ${expectedNumberOfEpisodes}`,
+                  );
+                  if (expectedNumberOfEpisodes > 1) {
+                    cy.get("ul[class*='css-1ddpce6-StyledEpisodeList']").should(
+                      'exist',
+                    );
+
+                    cy.get("ul[class*='css-1ddpce6-StyledEpisodeList']").within(
```
Comment (RichardPK) [inline cypress/integration/pages/onDemandTV/tests.js:90]: As above with the selectors being too specific, and to investigate whether a lot of this is shared logic between TV/radio and could be separated into its own test file?

File: cypress/integration/pages/onDemandRadio/tests.js (hunk 7 above)
Comment (LilyL0u) [reply cypress/integration/pages/onDemandRadio/tests.js:107]: This will be changed when we have the data-e2e tags put into the tv component. They are not there yet.

Comment (RichardPK) [reply cypress/integration/pages/onDemandRadio/tests.js:107]: So regardless of whether we are able to use a data-e2e value for these - using the emotion labels (`css-1sel12u` & `emzt7w80`) in the selector turns this into a bit of a time bomb (_drama_). 

When the component has its styles changed (either in Simorgh, or Psammead) - these will be regenerated, and the class selector will no longer find the element.

There are examples of using selectors without data-e2e attributes elsewhere in Simorgh, eg: https://github.com/bbc/simorgh/blob/e470060cd6a1853cd58cb18435f02c653de097bc/src%2Fapp%2Fcontainers%2FCpsFeaturesAnalysis%2Findex.test.jsx#L92

So even if we don't have data-e2e values we can use, we can still use other selectors, without including the emotion-assigned labels.
